patches/server/0807-Add-PlayerShieldDisableEvent.patch

```diff
@@ -0,0 +1,75 @@
+--- a/src/main/java/net/minecraft/world/entity/Mob.java
++++ b/src/main/java/net/minecraft/world/entity/Mob.java
+@@ -1588,9 +1588,16 @@ public abstract class Mob extends LivingEntity {
+             float f = 0.25F + (float) EnchantmentHelper.getBlockEfficiency(this) * 0.05F;
+ 
+             if (this.random.nextFloat() < f) {
++                // Paper start
++                io.papermc.paper.event.player.PlayerShieldDisableEvent event = new io.papermc.paper.event.player.PlayerShieldDisableEvent((org.bukkit.entity.Player) player.getBukkitEntity(), getBukkitEntity());
++                event.callEvent();
++
++                if (!event.isCancelled()) {
+                 player.getCooldowns().addCooldown(Items.SHIELD, 100);
+                 this.level.broadcastEntityEvent(player, (byte) 30);
+             }
++                // Paper end
++            }
+         }
+ 
+     }
+diff --git a/src/main/java/net/minecraft/world/entity/player/Player.java b/src/main/java/net/minecraft/world/entity/player/Player.java
+index cea92f1dc663bf0648b2bd877d86ca380a517bc9..450fe792114e40d6fbe7ba79ad4d419d36f3fe6b 100644
+--- a/src/main/java/net/minecraft/world/entity/player/Player.java
++++ b/src/main/java/net/minecraft/world/entity/player/Player.java
+@@ -936,7 +936,7 @@ public abstract class Player extends LivingEntity {
+     protected void blockUsingShield(LivingEntity attacker) {
+         super.blockUsingShield(attacker);
+         if (attacker.getMainHandItem().getItem() instanceof AxeItem) {
+-            this.disableShield(true);
++            this.disableShield(true, attacker); // Paper
+         }
+ 
+     }
+@@ -1403,6 +1403,11 @@ public abstract class Player extends LivingEntity {
+     }
+ 
+     public void disableShield(boolean sprinting) {
++        //Paper start
++        disableShield(sprinting, null);
++    }
++
++    public void disableShield(boolean sprinting, @Nullable LivingEntity attacker) {
+         float f = 0.25F + (float) EnchantmentHelper.getBlockEfficiency(this) * 0.05F;
+ 
+         if (sprinting) {
+@@ -1410,12 +1415,18 @@ public abstract class Player extends LivingEntity {
+         }
+ 
+         if (this.random.nextFloat() < f) {
+-            this.getCooldowns().addCooldown(Items.SHIELD, 100);
+-            this.stopUsingItem();
+-            this.level.broadcastEntityEvent(this, (byte) 30);
++            org.bukkit.entity.Entity finalAttacker = attacker != null ? attacker.getBukkitEntity() : null;
++            io.papermc.paper.event.player.PlayerShieldDisableEvent event = new io.papermc.paper.event.player.PlayerShieldDisableEvent((org.bukkit.entity.Player) getBukkitEntity(), finalAttacker);
++            event.callEvent();
++
++            if (!event.isCancelled()) {
++                this.getCooldowns().addCooldown(Items.SHIELD, 100);
++                this.stopUsingItem();
++                this.level.broadcastEntityEvent(this, (byte) 30);
++            }
+         }
+-
+     }
++    // Paper end
```

Comment:
This block of diff can be dramatically simplified. Once again, use the if (!event.callEvent) to combine the call and isCancelled check. Also, the lines that are only called if the event isn't cancelled dont actually change. So ideally those should not be included in the diff at all. I think this bit can be simplified to exactly 2 line insertions, 0 deletions. Since you don't need a reference to the event later, you can just construct it, call it, and check if its cancelled all in the exact same line. And the other line is the local var for the finalAttacker.